src/backend/stream.rs

```diff
@@ -0,0 +1,821 @@
+            debug_assert!(r == 0);
+
+            if (got as usize) < size / frame_size {
+                let mut latency: pa_usec_t = 0;
+                let rr: i32 = pa_stream_get_latency(s, &mut latency, ptr::null_mut());
+                if rr == -(PA_ERR_NODATA as i32) {
+                    /* this needs a better guess. */
+                    latency = 100 * PA_USEC_PER_MSEC;
+                }
+                debug_assert!(r == 0 || r == -(PA_ERR_NODATA as i32));
+                /* pa_stream_drain is useless, see PA bug# 866. this is a workaround. */
+                /* arbitrary safety margin: double the current latency. */
+                debug_assert!(self.drain_timer.is_null());
+                self.drain_timer =
+                    pa_context_rttime_new(self.context.context,
+                                          pa_rtclock_now() + 2 * latency,
+                                          Some(stream_drain_callback),
+                                          self as *const _ as *mut _);
+                self.shutdown = true;
+                return;
+            }
+
+            towrite -= size;
+        }
+
+        debug_assert!(towrite == 0);
+    }
+}
+
+unsafe extern fn stream_success_callback(_s: *mut pa_stream, _success: i32, u: *mut c_void)
+{
+  let stm = &*(u as *mut Stream);
+  pa_threaded_mainloop_signal(stm.context.mainloop, 0);
+}
+
+unsafe extern fn stream_drain_callback(a: *mut pa_mainloop_api, e: *mut pa_time_event, _tv: *const Struct_timeval, u: *mut c_void)
+{
+  let mut stm = &mut *(u as *mut Stream);
+  debug_assert!(stm.drain_timer == e);
+  stm.state_change_callback(cubeb::State::Drained);
+  /* there's no pa_rttime_free, so use this instead. */
+  (*a).time_free.unwrap()(stm.drain_timer);
+  stm.drain_timer = ptr::null_mut();
+  pa_threaded_mainloop_signal(stm.context.mainloop, 0);
+}
+
+unsafe extern fn stream_state_callback(s: *mut pa_stream,
+                                       u: *mut c_void)
+{
+    let stm = &mut *(u as *mut Stream);
+    if !PA_STREAM_IS_GOOD(pa_stream_get_state(s)) {
+        stm.state_change_callback(cubeb::State::Error);
+    }
+    pa_threaded_mainloop_signal(stm.context.mainloop, 0);
+}
+
+
+fn read_from_input(s: *mut pa_stream, buffer: *mut *const c_void, size: *mut usize) -> i32
+{
+  let readable_size = pa_stream_readable_size(s);
+  if readable_size > 0 {
+    if pa_stream_peek(s, buffer, size) < 0 {
+      return -1;
+    }
+  }
+
+  readable_size as i32
+}
+
+unsafe extern fn stream_write_callback(s: *mut pa_stream, nbytes: usize, u: *mut c_void)
+{
+//  LOGV("Output callback to be written buffer size %zd", nbytes);
+    let mut stm = &mut *(u as *mut Stream);
+    if stm.shutdown || stm.state != cubeb::State::Started {
+        return;
+    }
+
+    if stm.input_stream.is_null() {
+        // Output/playback only operation.
+        // Write directly to output
+        debug_assert!(!!stm.input_stream.is_null() && !stm.output_stream.is_null());
```

Comment:
Checking input_stream.is_null feels a bit suspicious since the `if stm.input_stream.is_null() {`